pkg/util/crypt/crypt_dev.go

```diff
@@ -0,0 +1,306 @@
+		sylog.Debugf("Error acquiring size of %s", path)
+		return "", "", err
+	}
+
+	squashFsSize := squashF.Size()
+
+	// Create a temporary file to format with crypt header
+	cryptF, err := ioutil.TempFile("", "crypt-")
+	if err != nil {
+		sylog.Debugf("Error creating temporary crypt file")
+		return "", "", err
+	}
+	defer cryptF.Close()
+
+	// Truncate the file taking the squashfs size and crypt header into account
+	// Crypt header is around 2MB in size. Slightly over-allocate to be safe
+	size := squashFsSize + 4*1024*1024 // 4MB for LUKS header
+
+	err = os.Truncate(cryptF.Name(), size)
+	if err != nil {
+		sylog.Debugf("Unable to truncate crypt file to size %d", size)
+		return "", "", err
+	}
+
+	// NOTE: This routine runs with root privileges. It's not necessary
+	// to explicitly set cmd's uid or gid here
+	cmd := exec.Command("/sbin/cryptsetup", "luksFormat", cryptF.Name())
+	stdin, err := cmd.StdinPipe()
+
+	if err != nil {
+		return "", "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, input)
+	}()
+
+	err = cmd.Run()
+	if err != nil {
+		sylog.Verbosef("Unable to format crypt device")
+		return "", "", err
+	}
+
+	// Associate the temporary crypt file with a loop device
+	loop, err := createLoop(cryptF, 0, uint64(size))
+	if err != nil {
+		return "", "", err
+	}
+
+	sp := strings.Split(loop, "/")
+	loopdev := sp[len(sp)-1]
+
+	fd, err := lock.Exclusive("/dev/mapper")
+	if err != nil {
+		sylog.Debugf("Unable to acquire lock on /dev/mapper")
+		return "", "", err
+	}
+	nextCrypt := getNextAvailableCryptDevice()
+	cmd = exec.Command("/sbin/cryptsetup", "luksOpen", loopdev, nextCrypt)
+	cmd.Dir = "/dev"
+	stdin, err = cmd.StdinPipe()
+	if err != nil {
+		return "", "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, input)
+	}()
+
+	err = cmd.Run()
+	if err != nil {
+		sylog.Verbosef("Unable to open crypt device: %s", nextCrypt)
+		return "", "", err
+	}
+
+	err = lock.Release(fd)
+	if err != nil {
+		sylog.Debugf("Unable to release lock on /dev/mapper")
+		return "", "", err
+	}
+
+	copyDeviceContents(path, "/dev/mapper/"+nextCrypt, squashFsSize)
+
+	// Open a new Temp file to stash the loop contents
+	loopF, err := ioutil.TempFile("", "loop-")
+	if err != nil {
+		sylog.Debugf("Error creating temporary crypt file")
+		return "", "", err
+	}
+
+	copyDeviceContents("/dev/"+loopdev, loopF.Name(), size)
+
+	return loopF.Name(), nextCrypt, err
+}
+
+// copyDeviceContents copies the contents of source to destination.
+// source and dest can either be a file or a block device
+func copyDeviceContents(source string, dest string, size int64) error {
+
+	squashFd, err := syscall.Open(source, syscall.O_RDWR, 0777)
+	if err != nil {
+		sylog.Debugf("Unable to open the file %s", source)
+		return err
+	}
+	defer syscall.Close(squashFd)
+
+	cryptFd, err := syscall.Open(dest, syscall.O_RDWR, 0777)
+
+	if err != nil {
+		sylog.Debugf("Unable to open the file: %s", dest)
+		return err
```

Comment:
Again, this should probably return some debugging with the error message:

```suggestion
		return fmt.Errorf("unable to open file: %s: %v", dest, err)
```